llvm/lib/Target/AMDGPU/SIFixSGPRCopies.cpp

```diff
@@ -1086,10 +1086,22 @@ void SIFixSGPRCopies::lowerVGPR2SGPRCopies(MachineFunction &MF) {
         TRI->getRegClassForOperandReg(*MRI, MI->getOperand(1));
     size_t SrcSize = TRI->getRegSizeInBits(*SrcRC);
     if (SrcSize == 16) {
-      // HACK to handle possible 16bit VGPR source
-      auto MIB = BuildMI(*MBB, MI, MI->getDebugLoc(),
-                         TII->get(AMDGPU::V_READFIRSTLANE_B32), DstReg);
-      MIB.addReg(SrcReg, 0, AMDGPU::NoSubRegister);
+      assert(MF.getSubtarget<GCNSubtarget>().useRealTrue16Insts() &&
+             "We do not expect to see 16-bit copies from VGPR to SGPR unless "
+             "we have 16-bit VGPRs");
+      assert(MRI->getRegClass(DstReg) == &AMDGPU::SGPR_LO16RegClass ||
+             MRI->getRegClass(DstReg) == &AMDGPU::SReg_32RegClass ||
+             MRI->getRegClass(DstReg) == &AMDGPU::SReg_32_XM0RegClass);
+      // There is no V_READFIRSTLANE_B16, so legalize the dst/src reg to 32 bits
+      MRI->setRegClass(DstReg, &AMDGPU::SReg_32_XM0RegClass);
+      Register VReg32 = MRI->createVirtualRegister(&AMDGPU::VGPR_32RegClass);
+      const DebugLoc &DL = MI->getDebugLoc();
+      BuildMI(*MBB, MI, DL, TII->get(TargetOpcode::SUBREG_TO_REG), VReg32)
+          .addImm(0)
+          .addReg(SrcReg, 0)
+          .addImm(AMDGPU::lo16);
```

Comment:
SUBREG_TO_REG is a malformed operation and we should not use it. Either use an IMPLICIT_DEF + an INSERT_SUBREG or REG_SEQUENCE

Reply:
Hi Matt. Why do you think SUBREG_TO_REG is malformed? It seems it is recognizable by many codegen passes. Is it because we don't have proper support for it?

We have some other places that use this opcode for true16 flow and we might need to update them together if this is not the right one.

Reply:
No, as in the operation is fundamentally broken as a representation and should be deleted from the IR. There are 2 issues, first here you are asserting the high bits are 0 when they are not in use here. Second, the problem is it's introducing a hidden dependence on the unnamed super register component which is no good. The coalescer has to insert hacky super register defs to preserve the other subregisters which were unnamed prior to coalescing

Reply:
I see. Thanks! Updated the code